### PR TITLE
[#49] Add optional separate channel for non-error outcomes

### PR DESCRIPTION
**Pull request by z-cran**
### Ticket
https://github.com/massenz/go-statemachine/issues/49

### Summary of Changes
- Added an optional `outcomes` channel if a `notifications` channel is also provided
  - If `outcomes` is provided, it will receive OK outcomes; `notifications` will receive every other outcome 

### Testing
- Added new unit tests for outcomes & confirmed existing tests pass

### CODE REVIEW

**Review by massenz — CHANGES_REQUESTED** [review]

**Comment by z-cran** [comment]
Thanks for the feedback! I moved the logic out of `listener` and into `sub_pub` as recommended and incorporated your other notes (such as in `main` and `entrypoint`).

One note: for the tests, I found I had to add a small sleep between adding an item to a channel & checking the message on the queue. Without this, tests were failing intermittently for me.

**Comment by z-cran** [comment]
After speaking with @YulyTenorio , we also were hoping to add the capability to provide an empty `-acks` if we did not want Ok outcomes to be published. The PR was updated to incorporate this.

File: README.md [hunk 1]
```diff
@@ -453,6 +453,8 @@ Usage of build/bin/sm-server:
     	Max number of attempts for a recoverable error to be retried against the Redis cluster (default 3)
   -notifications string
     	The name of the notification topic in SQS to publish events' outcomes to; if not specified, no outcomes will be published
+  -acks string
```
Comment (massenz) [inline README.md:456]: this is a nit, but currently the message here and the one in the app do not match

File: cmd/main.go [hunk 2]
```diff
@@ -80,13 +80,21 @@ func main() {
 	var notificationsTopic = flag.String("notifications", "",
 		"The name of the notification topic in SQS to publish events' outcomes to; if not "+
 			"specified, no outcomes will be published")
+	var acksTopic = flag.String("acks", "",
+		"If `acks` is specified and provided a value, all positive outcomes (Ok) will be published here, and only "+
+			"errors will be published to the `notifications` queue. If specified and not provided a value, "+
+			"Ok outcomes will not be published. If not specified, all results "+
```
Comment (massenz) [inline cmd/main.go:86]: I don't actually think you can achieve the "specified and not provided a value"?
something like `-ack -foo bar` will assume that the value of `ack` is `-foo` (and the argument `-foo` if required, is missing)

Haven't really tried it, though - have you?
At any rate, I don't like this pattern, as it's confusing.

Please add a boolean `-notify-errors-only` to achieve the same outcome instead (and add a check that it cannot be specified alongside `-ack`)

File: cmd/main.go [hunk 3]
```diff
@@ -80,13 +80,21 @@ func main() {
...
+			"will published to the `-notifications` queue regardless of their error status")
```
Comment (massenz) [inline cmd/main.go:87]: nit: `will be published`

File: cmd/main.go [hunk 4]
```diff
@@ -154,6 +165,17 @@ func main() {
 	logger.Fatal(srv.ListenAndServe())
 }
 
+// checkForFlag checks whether a flag was provided
+func checkForFlag(name string) bool {
```
Comment (massenz) [inline cmd/main.go:169]: this makes the logic really convoluted and very easy to "use it wrong"

File: pubsub/sqs_pub.go [hunk 5]
```diff
@@ -56,13 +57,17 @@ func GetQueueUrl(client *sqs.SQS, topic string) string {
 	return *out.QueueUrl
 }
 
-// Publish sends an error message to the DLQ `topic`
-func (s *SqsPublisher) Publish(topic string) {
-	queueUrl := GetQueueUrl(s.client, topic)
-	s.logger = log.NewLog(fmt.Sprintf("SQS-Pub{%s}", topic))
-	s.logger.Info("SQS Publisher started for queue: %s", queueUrl)
-	for eventResponse := range s.errors {
+// Publish sends an error message to provided topics
```
Comment (massenz) [inline pubsub/sqs_pub.go:60]: can you please update the docstring to actually reflect what Publish does?

File: pubsub/sqs_pub.go [hunk 6]
```diff
@@ -56,13 +57,17 @@ func GetQueueUrl(client *sqs.SQS, topic string) string {
...
+func (s *SqsPublisher) Publish(errorsTopic string, acksTopic string) {
+	s.logger.Info("SQS Publisher started for topics: %s %s", errorsTopic, acksTopic)
+	for eventResponse := range s.notifications {
 		delay := int64(0)
+
+		queueUrl := GetQueueUrl(s.client, errorsTopic)
```
Comment (massenz) [inline pubsub/sqs_pub.go:66]: please really, really, don't do this
you're looking up the URL for your topic (and it's a full round-trip to AWS management APIs) for *every* message

Please look up and store the queue URL in TWO variables **outside** the `for` loop, and then pick the right one based on (a) is it an OK message and (b) are we publishing Ok messages

File: pubsub/sqs_pub.go [hunk 7]
```diff
@@ -56,13 +57,17 @@ func GetQueueUrl(client *sqs.SQS, topic string) string {
...
+		if acksTopic != "" && !s.ignoreOkOutcomes && eventResponse.Outcome.Code == protos.EventOutcome_Ok {
```
Comment (massenz) [inline pubsub/sqs_pub.go:67]: not to mention, with this logic, you're still publishin on Ok - just to the notification topic (unless I'm missing something)

File: pubsub/sqs_pub_test.go [hunk 8]
```diff
@@ -50,9 +51,10 @@ var _ = Describe("SQS Publisher", func() {
 			done := make(chan interface{})
 			go func() {
 				defer close(done)
-				go testPublisher.Publish(getQueueName(notificationsQueue))
+				go testPublisher.Publish(getQueueName(notificationsQueue), "")
 			}()
 			notificationsCh <- notification
+			time.Sleep(channelWait)
```
Comment (massenz) [inline pubsub/sqs_pub_test.go:57]: don't do `Sleep` in tests - the pattern to follow is the one used in the other tests, such as in https://github.com/massenz/go-statemachine/blob/main/pubsub/sqs_pub_test.go#L100-L112

File: pubsub/sqs_pub_test.go [hunk 9]
```diff
@@ -97,11 +100,52 @@ var _ = Describe("SQS Publisher", func() {
...
+			}
+			done := make(chan interface{})
+			go func() {
+				defer close(done)
+				go testPublisher.Publish(getQueueName(notificationsQueue), getQueueName(acksQueue))
+			}()
+			var response protos.EventResponse
+
+			notificationsCh <- notification
+			time.Sleep(channelWait)
```
Comment (massenz) [inline pubsub/sqs_pub_test.go:124]: here too

File: pubsub/sqs_pub_test.go [hunk 10]
```diff
@@ -157,5 +202,33 @@ var _ = Describe("SQS Publisher", func() {
...
+			testPublisher = pubsub.NewSqsPublisher(notificationsCh, &sqsUrl, true)
+			done := make(chan interface{})
+			go func() {
+				defer close(done)
+				go testPublisher.Publish(getQueueName(notificationsQueue), getQueueName(acksQueue))
+			}()
+
+			notificationsCh <- ack
+			time.Sleep(channelWait)
+			res := getSqsMessage(getQueueName(acksQueue))
```
Comment (massenz) [inline pubsub/sqs_pub_test.go:221]: out of curiousity, you may add a `getSqsMessage` for the `notificationsQueue` and confirm whether it gets the ok message (which I think it does with current code); you need to check them both anyway and confirm we drop those

**Review by massenz — CHANGES_REQUESTED** [review]

**Review by massenz — APPROVED** [review]

**Pull request merged**